### PR TITLE
Add test cases for multiple Referrer-Policy headers

### PR DESCRIPTION
**Pull request by estark37**
The test cases exercise various ways that multiple policies can be delivered in
Referrer-Policy headers, such as:

Referrer-Policy: origin, no-referrer

Referrer-Policy: origin
Referrer-Policy: no-referrer

Referrer-Policy: origin
Referrer-Policy: origin no-referrer

As requested in https://github.com/w3c/webappsec-referrer-policy/issues/81

### CODE REVIEW

**Comment by estark37** [comment]
Hoping for review from some combination of @mikewest @jeisinger @annevk -- thanks!

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/208339579)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 6a86d428706708d667f2e3d2d7d8d9d9251706b9
Using browser at version BuildID 20170306110339; SourceStamp 966464a68a2cb3ca1125808e34abb5c1d34e3797
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-and-values.html">/referrer-policy/generic/multiple-headers-and-values.html</a></summary>


|                            Subtest                            | Results |                                                                                           Messages                                                                                           |
|---------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                               | OK      |                                                                                                                                                                                              |
| `Image uses the last recognized Referrer-Policy header value` | FAIL    | `assert_equals: Referrer inside 'srcdoc' without its own policy should be the same as embedder's referrer. expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-combined.html">/referrer-policy/generic/multiple-headers-combined.html</a></summary>


|                            Subtest                            | Results |                                                                                           Messages                                                                                           |
|---------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                               | OK      |                                                                                                                                                                                              |
| `Image uses the last recognized Referrer-Policy header value` | FAIL    | `assert_equals: Referrer inside 'srcdoc' without its own policy should be the same as embedder's referrer. expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-one-invalid.html">/referrer-policy/generic/multiple-headers-one-invalid.html</a></summary>


|                            Subtest                            | Results |                                                                                           Messages                                                                                           |
|---------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                               | OK      |                                                                                                                                                                                              |
| `Image uses the last recognized valid Referrer-Policy header` | FAIL    | `assert_equals: Referrer inside 'srcdoc' without its own policy should be the same as embedder's referrer. expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers.html">/referrer-policy/generic/multiple-headers.html</a></summary>


|                         Subtest                         | Results |                                                                                           Messages                                                                                           |
|---------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                         | OK      |                                                                                                                                                                                              |
| `Image uses the last recognized Referrer-Policy header` | FAIL    | `assert_equals: Referrer inside 'srcdoc' without its own policy should be the same as embedder's referrer. expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/208339580)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 6a86d428706708d667f2e3d2d7d8d9d9251706b9
Using browser at version 58.0.3026.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-and-values.html">/referrer-policy/generic/multiple-headers-and-values.html</a></summary>


|                            Subtest                            | Results | Messages |
|---------------------------------------------------------------|---------|----------|
|                                                               | OK      |          |
| `Image uses the last recognized Referrer-Policy header value` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-combined.html">/referrer-policy/generic/multiple-headers-combined.html</a></summary>


|                            Subtest                            | Results |                                              Messages                                             |
|---------------------------------------------------------------|---------|---------------------------------------------------------------------------------------------------|
|                                                               | OK      |                                                                                                   |
| `Image uses the last recognized Referrer-Policy header value` | FAIL    | `assert_equals: expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers-one-invalid.html">/referrer-policy/generic/multiple-headers-one-invalid.html</a></summary>


|                            Subtest                            | Results | Messages |
|---------------------------------------------------------------|---------|----------|
|                                                               | OK      |          |
| `Image uses the last recognized valid Referrer-Policy header` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5054/referrer-policy/generic/multiple-headers.html">/referrer-policy/generic/multiple-headers.html</a></summary>


|                         Subtest                         | Results |                                              Messages                                             |
|---------------------------------------------------------|---------|---------------------------------------------------------------------------------------------------|
|                                                         | OK      |                                                                                                   |
| `Image uses the last recognized Referrer-Policy header` | FAIL    | `assert_equals: expected (string) "http://web-platform.test:8000/" but got (undefined) undefined` |

</details>
</details>

**Comment by estark37** [comment]
(note that these tests currently fail in Chrome because of https://crbug.com/698813, with a fix out for review in https://codereview.chromium.org/2733943002/)

**Review by jeisinger — APPROVED** [review]

**Comment by annevk** [comment]
So technically, if you have an unrecognized policy-token, everything is supposed to be dropped, because the given value ends up not matching the ABNF.

So given that I think some of the expectations here are wrong. E.g., for the invalid case I would expect the result to be the same as if no header was transmitted.

**Comment by annevk** [comment]
To be clear, that might mean we want to change something in the Referrer Policy specification instead, to make the parsing of header values a bit more lenient so it's easier to introduce new values going forward.

**Comment by estark37** [comment]
Spec has been updated so that these expectations are correct now, I think

File: referrer-policy/generic/multiple-headers-one-invalid.html
Comment (annevk) [inline referrer-policy/generic/multiple-headers-one-invalid.html]: If one of the headers is invalid, that means it fails overall so this shouldn't assume origin I think. This should assume whatever is the default.

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5054/)

File: referrer-policy/generic/multiple-headers-one-invalid.html
Comment (estark37) [inline referrer-policy/generic/multiple-headers-one-invalid.html]: Ah, right. Updated this test, added another one for an unknown token (multiple-headers-one-unknown-token).

Comment (estark37) [inline referrer-policy/generic/multiple-headers-one-invalid.html]: (Also, I have to update the Chrome implementation to match this behavior: https://crbug.com/705950)

**Comment by annevk** [comment]
LGTM, assuming the basics are okay in terms of the tests running and such.

**Review by mikewest — APPROVED** [review]

**Comment by zcorpan** [comment]
Should this be merged?

**Review by domfarolino — APPROVED** [review]
LGTM. Will defer to @estark37 or @jeisinger to merge.

**Comment by mikewest** [comment]
Emily is OOO; I'll merge it for her and see what explodes. :)

**Pull request merged**
